lib/services/embeddings.go

```diff
@@ -0,0 +1,312 @@
+	NodeEmbeddingWatcherConfig
+
+	// initializationC is used to check whether the initial sync has completed
+	initializationC chan struct{}
+	once            sync.Once
+
+	currentNodes map[string]*embeddedNode
+
+	rw    sync.RWMutex
+	stale atomic.Bool
+}
+
+type embeddedNode struct {
+	node           types.Server
+	needsEmbedding bool
+	// Last update allows to avoid most race conditions
+	// Before updating or deleting the node, the caller must
+	// check if it has been edited in the meantime.
+	lastUpdate time.Time
+}
+
+func (e *embeddedNode) gotEmbedded() {
+	e.needsEmbedding = false
+}
+
+// resourceKind specifies the resource kind to watch.
+func (n *nodeEmbeddingCollector) resourceKind() string {
+	return types.KindNode
+}
+
+// getResourcesAndUpdateCurrent is called when the resources should be
+// (re-)fetched directly.
+func (n *nodeEmbeddingCollector) getResourcesAndUpdateCurrent(ctx context.Context) error {
+	timestamp := time.Now()
+	allNodes, err := n.getNodes(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	n.rw.RLock()
+	toRemove := make([]string, 0, len(n.currentNodes))
+	// If we knew a node which is not in the full node list anymore we drop
+	// it from the index
+	for _, knownNode := range n.currentNodes {
+		if _, ok := allNodes[knownNode.node.GetName()]; !ok {
+			toRemove = append(toRemove, knownNode.node.GetName())
+		}
+	}
+	n.rw.RUnlock()
+
+	n.AddNodes(allNodes, timestamp)
+	n.RemoveNodes(toRemove, timestamp)
+
+	n.defineCollectorAsInitialized()
+	n.stale.Store(false)
+	return nil
+}
+
+// AddNodes takes a map of new or updated nodes, stores them and flags them for
+// embedding.
+func (n *nodeEmbeddingCollector) AddNodes(nodes map[string]types.Server, timestamp time.Time) {
+	n.rw.Lock()
+	defer n.rw.Unlock()
+	for nodeName, node := range nodes {
+		// If the node is already known and has been edited in the meantime we
+		// don't want to override as we don't have the latest version
+		if currentNode, ok := n.currentNodes[nodeName]; ok {
+			if currentNode.lastUpdate.After(timestamp) {
+				continue
+			}
+		}
+		n.currentNodes[nodeName] = &embeddedNode{
+			node:           node,
+			needsEmbedding: true,
+			lastUpdate:     timestamp,
+		}
+	}
+}
+
+// RemoveNodes takes a list of node names, removes them from the vector index
+// and from the collector tracking.
+func (n *nodeEmbeddingCollector) RemoveNodes(nodeNames []string, timestamp time.Time) {
+	n.rw.Lock()
+	defer n.rw.Unlock()
+	for _, nodeName := range nodeNames {
+		if n.currentNodes[nodeName].lastUpdate.Before(timestamp) {
+			delete(n.currentNodes, nodeName)
+		}
+	}
+}
+
+// RunIndexation walks through all collector-tracked nodes and runs a batch
+// embedding on all nodes needing embeddings. The embeddings are then inserted
+// into the vector index. This process is ran asynchronously to reduce the load
+// and leverage OpenAI's batch embedding API.
+func (n *nodeEmbeddingCollector) RunIndexation(ctx context.Context) error {
```

Comment:
If I understand this correctly, if two goroutines run this method, both will read all nodes, then both will create embeddings and update the index with the same values. Shouldn't we just take the exclusive lock at the beginning and double-check if the embedding generation is needed to avoid it?

Reply:
The only sensitive part in this is when the full sync runs and the function getResourcesAndUpdateCurrent acquires the lock.

There's a single routine triggering the embeddings so we should not be doing double work, but even if this happens this does not affect consistency because of the timestamp check.